### ISSUE TITLE
token renewal operation failed due to timeout MSAL

### ISSUE THREAD

**Issue by ashishbhulani** [AzureAD/microsoft-authentication-library-for-js#1592]
Hi,

Library : "@azure/msal-angular": "^1.0.0-beta.5"
             "msal": "^1.2.2"

I have integrated MSAL library with my angular 8 application. Everything works fine except i keep getting an error token renewal operation failed due to timeout as soon as the token is expired. I was wondering how to fix this.

Below is the MSAL_config that i am using : 
```js
"auth": {
        
        "clientId": "xxxxx",
        "authority": "https://login.microsoftonline.com/xxx",
        "validateAuthority":"true",
        "postLogoutRedirectUri": "http://localhost:4200/",
        "navigateToLoginRequestUrl": true,
        "redirectUri":"http://localhost:4200/"
    },
    
    "scopes":["user.read", "openid", "profile"],
    "popUp": false,
    "unprotectedResources": ["https://www.microsoft.com/en-us/"],
    "protectedResourceMap":[["https://graph.microsoft.com/v1.0/me", ["user.read"]]],
    "system":{
        "loadFrameTimeout":10000
    },
    "cache": {
      "cacheLocation": "localStorage",
      "storeAuthStateInCookie":true

    }
```
I have created a new token interceptor which pull the token everytime a http request is amde. Below is the code, here i face issue when the token is expired and i get the error. Please help me to resolve this as it is affecting the project.
```js
if (this.authService.getAccount()) {
 

                let token: string;
                return from(
                    this.authService.acquireTokenSilent(this.loginRequest)
                        .then((response: AuthResponse) => {
            
                            token = response.idToken.rawIdToken;
                            const authHeader = `Bearer ${token}`;
                            if (!request.headers.has('Cache-Control')) {
                                request = request.clone({ headers: request.headers.set('Cache-Control', 'no-cache' + '') });
                            }

                            if (!request.headers.has('Pragma')) {
                                request = request.clone({ headers: request.headers.set('Pragma', 'no-cache' + '') });
                            }

                            if (!request.headers.has('Expires')) {
                                request = request.clone({ headers: request.headers.set('Expires', 'Sat, 01 Jan 2000 00:00:00 GMT' + '') });
                            }
                            return request.clone({
                                setHeaders: {
                                    Authorization: authHeader,
                                }
                            });

                        })
                )
                    .pipe(
                        mergeMap(nextReq => next.handle(nextReq)),
                        tap(
                            event => { },
                            err => {
                                if (err) {
                                    var iframes = document.querySelectorAll('iframe');
                                    for (var i = 0; i < iframes.length; i++) {
                                        iframes[i].parentNode.removeChild(iframes[i]);
                                    }
                                    debugger
                                    
                                    this.authService.handleRedirectCallback((err: AuthError, response) => {
                                        debugger
                                        this.authService.loginRedirect();
                                        if (err) {

                                            console.error('Redirect Error: ', err.errorMessage);
                                            return;
                                        }
                                        debugger
                                        this.authService.loginRedirect();
                                        console.log('Redirect Success: ', response);
                                    });
                                    this.broadcastService.broadcast('msal:notAuthorized', err.message);
                                }
                            }
                        )
                    );
            }
```
So in the above code it goes to the error but it never enters  this.authService.handleRedirectCallback can someone please help me with this. I need the token to be renewed.

**Comment by pkanher617** [comment]
@ashishbhulani Could you please use the template when creating an issue? Without knowing what library and version you are using it is difficult to understand your issue.

**Comment by ashishbhulani** [comment]
@pkanher617 How do i use the template?

**Comment by pkanher617** [comment]
@ashishbhulani when you are [creating a new issue](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/new/choose), you can choose one of the options to give you the relevant template.

It appears you are calling handleRedirectCallback in the wrong place. It needs to be set directly after the constructor of the authService, as it gets called when the page is loaded again with the auth information.

**Comment by ashishbhulani** [comment]
@pkanher617 can you help me modify the code what exactly i am doing wrong. I have written this code inside tojen interceptor service.

**Comment by pkanher617** [comment]
@ashishbhulani I just told you. Directly after the creating the authService object on page load, you need to call handleRedirectCallback. Currrently you are doing it after an error in acquireTokenSilent, which is incorrect usage. 

We have some [angular dev samples](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples) in the repo you can use for more information.

**Comment by ashishbhulani** [comment]
@pkanher617 Actually the issue is that once the token is getting expired i am unable to renew it and that is the reason why i have put in error. So when the token is not expired the acquiretokensilent helps me to authenticate with the api. I face the token renew issue only when it is expired. So can you help me to re-write this code or can you paste here an example on how to fix this in this situation. As i have already gone through the demo samples but i dont see the usage of this. I request you to help me with the code.

**Comment by ashishbhulani** [comment]
@jasonnutter Can you help me on this?

**Comment by ashishbhulani** [comment]
@pkanher617 @jasonnutter How do i renew my token. can you please help me with that?

**Comment by jasonnutter** [comment]
@ashishbhulani You should be able to handle this without modifying the interceptor. When you make the http request and you receive an error, if it a timeout or interaction required error, invoke acquireTokenPopup with the same set of scopes, and then make the request again. 

```js
const GRAPH_ENDPOINT = "https://graph.microsoft.com/v1.0/me";

this.http.get(GRAPH_ENDPOINT)
    .subscribe({
      next: (profile) => {
        this.profile = profile;
      },
      error: (err: AuthError) => {
          this.authService.acquireTokenPopup({
            scopes: this.authService.getScopesForEndpoint(GRAPH_ENDPOINT)
          })
          .then(() => {
            this.http.get(GRAPH_ENDPOINT)
              .toPromise()
              .then(profile => {
                this.profile = profile;
              });
          });
      }
    });

```

Note, if you want to use `acquireTokenRedirect` or `loginRedirect` instead, your application will need to implement `handleRedirectCallback` separately, not inside the interceptor or where you make the http request. Instead, it needs to be invoked on page load, [as demonstrated in the Angular 8 sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/angular8-sample-app/src/app/app.component.ts#L27). Remember, if you call redirect, the browser will fully redirect away from your application and lose all context, which is why we recommend acquireTokenPopup instead for this scenario.

**Comment by ashishbhulani** [comment]
@jasonnutter This worked but now when i acquiretokensilently i am getting the old token where as in local storage i see the new token.
this.authService.acquireTokenSilent(this.loginRequest)
                        .then((response: AuthResponse)

In the response i still see the old token what could be the reason?

**Comment by ashishbhulani** [comment]
@jasonnutter can you tell what is the actual issue here?

**Comment by ashishbhulani** [comment]
@jasonnutter can you please help me. I am stuck with renewal of token?

**Comment by jasonnutter** [comment]
> @jasonnutter This worked but now when i acquiretokensilently i am getting the old token where as in local storage i see the new token.
> this.authService.acquireTokenSilent(this.loginRequest)
> .then((response: AuthResponse)
> 
> In the response i still see the old token what could be the reason?

We've heard some other reports of this happening, I'll follow up if I need more information.

**Comment by eeskildsen** [comment]
@jasonnutter I'm getting the same error. msal 1.3.1., msal-angular 1.0.0.

Auth only works right after I clear the browser cache. Hours later:

    ERROR Error: Uncaught (in promise): ClientAuthError: URL navigated to is https://login.microsoftonline.com/<TenantId>/oauth2/v2.0/authorize?response_type=id_token&scope=openid%20profile&client_id=<ClientId>&redirect_uri=<RedirectUri>&state=<State>&nonce=<Nonce>&client_info=1&x-client-SKU=MSAL.JS&x-client-Ver=1.3.1&login_hint=<MyUsername>&client-request-id=<RequestId>&prompt=none&response_mode=fragment, Token renewal operation failed due to timeout.

Here's my code. There are 3 places where I'm configuring/calling MSAL. I don't *think* I'm doing anything fancy. I'm basing it on the [Angular 8 sample app](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/angular8-sample-app/).

1. `app.module.ts` -> `NgModule` -> `imports`:

       MsalModule.forRoot({
               auth: {
                   clientId: "<ClientId>",
                   authority: "https://login.microsoftonline.com/<TenantId>/",
                   validateAuthority: true,
                   redirectUri: '<SiteRoot>',
                   postLogoutRedirectUri: "<SomeUrl>",
                   navigateToLoginRequestUrl: true,        
               },
               cache: {
                   cacheLocation : "localStorage",
                   storeAuthStateInCookie: isIE
               },
               framework: {
                   unprotectedResources: ['https://www.microsoft.com/en-us/'],
               }
           },
           {
               popUp: !isIE,
               consentScopes: [ "https://graph.microsoft.com/User.Read", "api://<ApiId>/API" ],
               extraQueryParameters: {},
               protectedResourceMap: [
                   ['<SiteRoot>', ['api://<ApiId>/API']],
                   ['https://graph.microsoft.com', ['user.read', 'openid', 'profile', 'email', 'offline_access']]
               ]
           })

1. `app.module.ts` -> `NgModule` -> `providers`

       { provide: HTTP_INTERCEPTORS, useClass: MsalInterceptor, multi: true }

1. `app.component.ts` -> `AppComponent` -> `ngOnInit`

       ngOnInit() {
           this.authService.handleRedirectCallback((authError, response) => {
               if (authError) {
                   console.error('Redirect Error: ', authError.errorMessage);
                   return;
               }
       
               console.log('Redirect Success: ', response);
           });
       }

**Comment by ryandegruyter** [comment]
@jasonnutter  We're also experience a token renewal problem. We have our handleRedirectCallback setup in the root component:
`
    this.msalService.handleRedirectCallback((authError, response) => {
      this.authError = authError;
      if (authError && !this.isLoginInProgress()) {
        this.router.navigate(['login-failed']);
      }
    });`

Error:

ERROR ClientAuthError: URL navigated to is https://login.microsoftonline.com/c89214f1-7515-48ee-9cd0-9b859ed3e4c4/oauth2/v2.0/authorize?response_type=id_token&scope=openid%20profile&client_id=xx-1916-43a3-bdab-a808a9fc29cd&redirect_uri=http%3A%2F%2Flocalhost%3A4200&state=$$OUR_TOKEN_REMOVED$$&nonce=ba4d0a17-xxxx-4112-b997-5d79f8d483db&client_info=1&x-client-SKU=MSAL.JS&x-client-Ver=1.3.1&login_hint=JE34524%40ACC-xxx-xxx.COM&client-request-id=fa6cc7f8-xxxx-474d-aa05-b256ffa7b29c&prompt=none&response_mode=fragment, Token renewal operation failed due to timeout.
    at ClientAuthError.AuthError [as constructor] (http://localhost:4200/vendor.js:190343:28)
    at new ClientAuthError (http://localhost:4200/vendor.js:190481:28)
    at Function.push../node_modules/msal/lib-es6/error/ClientAuthError.js.ClientAuthError.createTokenRenewalTimeoutError


Also why doesn't the callback handle the error?

**Comment by ryandegruyter** [comment]
Also, a hard refresh of the page and token retrieval works.

**Comment by eeskildsen** [comment]
As a workaround, I set `loadFrameTimeout` to `30000`. That seemed to let me through. It's also possible I accidentally cleared my cache at some point. I'll double-check tomorrow.

    MsalModule.forRoot({
        system: {
            loadFrameTimeout: 30000
        },
        ...

**Comment by ryandegruyter** [comment]
Nope not working here with timeout:

Wed, 17 Jun 2020 15:09:30 GMT:1.3.1-Verbose Loading frame has timed out after: 30 seconds for scope/authority 7e20f2ad-1916-43a3-bdab-a808a9fc29cd|**undefined**:eyJpZCI6ImQwNDU1YjllLWE0NTItNDIwMS1hNDE0LWIwMTgyMGFmNDMzMiIsInRzIjoxNTxxxxdGlvbiJ9 | Level: 3 | ContainsPii: false

Something undefined here though.

@jasonnutter if you have any leads for us? This is deployed to more than then thousand of users and we have no pattern. Some users experience it, others not...

**Comment by ryandegruyter** [comment]
I noticed this request fails when it happens:

```
curl 'https://login.microsoftonline.com/c89214f1-7515-48ee-xxxx-9b859ed3e4c4/oauth2/v2.0/authorize?response_type=id_token&scope=openid%20profile&client_id=7e20f2ad-1916-43a3-bdab-a808a9fc29cd&redirect_uri=http%3A%2F%2Flocalhost%3A4200&state=eyJpZCI6IjlmZjgzYzY0LTk1ZTktNGUzOS05YjI0xxxxxkyNDY0Mzc0LCJtZXRob2QiOiJzaWxlbnRJbnRlcmFjdGlvbiJ9&nonce=2fb4194f-xxx-4f74-b14d-968c2304ed3d&client_info=1&x-client-SKU=MSAL.JS&x-client-Ver=1.3.1&login_hint=JE26806%40ACC-xxx-xxxx.COM&client-request-id=6ee72ee5-xxx-42bd-9028-21aa5eaa619f&prompt=none&response_mode=fragment'
```

**Comment by patrasm** [comment]
I was fighting with this issue for the past week.  I ended up implementing a solution similar to this:  

https://github.com/Azure-Samples/active-directory-javascript-graphapi-v2/blob/60e53cf861c0a51a690aff8043032dd94f1b488d/JavaScriptSPA/authRedirect.js

**Comment by m-mattu** [comment]
Experiencing this issue as well. The constant that I have observed is it only occurs on Chrome Browsers (Desktop & mobile). However when I test the same app in Microsoft Edge (83.0.478.56) or Firefox the authentication works. I have Chrome configured to allow all cookies.

Here is my msal `config:
`export const authConfig: Configuration = {
  auth: {
    clientId: appId,
    authority: "<authority>",
    validateAuthority: true,
    redirectUri: redirectUri,
    postLogoutRedirectUri: "https://account.activedirectory.windowsazure.com",
    navigateToLoginRequestUrl: false,
  },
  cache: {
    cacheLocation: "sessionStorage",
    storeAuthStateInCookie: true,
  },
  system: {
    loadFrameTimeout: 10000,
    logger: new Logger(loggerCallback, {
      level: LogLevel.Info,
      piiLoggingEnabled: false,
    }),
  },
};`

And using React Hooks I have the redirectCallback registered on App mount:
`useEffect(() => {
    //register MSAL Redirect Callback
    msalAuthClient.handleRedirectCallback(async (error: any, response: any) => {
      if (error) {
        const errorMessage = error.errorMessage
          ? error.errorMessage
          : "Unable to acquire acces token";
        logger.error({
          message: errorMessage,
          metaData: error,
        });
        history.push("/login");
        return;
      } else {
        //login to ERP
        try {
          const account = msalAuthClient.getAccount();
          ai.appInsights?.setAuthenticatedUserContext(
            account.userName,
            account.accountIdentifier,
            true
          );
          const msalToken = response.idToken.rawIdToken;
          const accessToken = await MsalAuthService.getAccessToken(USER_SCOPE);
          const erpToken = await erpAuth.exchangeMsalToken(msalToken);
          dispatch(setJdeToken(jdeToken.userInfo.token));
          dispatch(setAadUser(account));
          history.push("/");
        } catch (err) {
          const error = normalizeError(err);
          logger.error({
            message: `Error logging in to JD Edwards: ${error.message}`,
            metaData: error.errorProps,
          });
          history.push("/login");
        }
      }
    });
  }, []);`

I followed this guide: https://docs.microsoft.com/en-us/graph/tutorials/react

**Comment by anth-git** [comment]
I'm still encountering "_Token renewal operation failed due to timeout_" error in v1.3.3.
It fails roughly 1 in 4 login attempts.

When it fails:
- **broadcastService** broadcasts _'msal:loginSuccess'_ event
- **msalService.getAccount()** returns valid account object
- **msalGuard.canActivate()** returns true

but at the same time:

- **msalService.acquireTokenSilent(...)** throws timeout error
- any call to the API (this.http.get(...)) throws timeout error (I guess from the **MsalInterceptor**)

**Comment by abellisai** [comment]
Me too, i'm still encountering this error when MSAL rise "monitorIframeForHash unable to find hash in url, timing out"

**Comment by llakhani** [comment]
Yes, I'm facing same issue - 'token renewal operation failed due to timeout token renewal failed'. Any update ?

**Comment by ZacharyHiggins-dbMotion** [comment]
This is occurring for our project as well. Is there any additional information we can provide? Increasing the timeout does not resolve the issue. Logging in is fine, but fails when calling our function app API from the site.

**Comment by anth-git** [comment]
I logged the value of the **href** variable in the **monitorIframeForHash** function.

When it works:
```
about:blank (several times)
http://localhost:4200/#state=eyJ…
```

When it doesn't work:
```
about:blank (several times)
http://localhost:4200/# (without anything after the #)
```

The problem might be with the POLLING_INTERVAL_MS. If I increase it from the default 50ms to 500ms I'm not able to sign in whatsoever. If I decrease it to 5ms, it seems to be working every time.

**Comment by ZacharyHiggins-dbMotion** [comment]
> I logged the value of the **href** variable in the **monitorIframeForHash** function.
> 
> When it works:
> 
> ```
> about:blank (several times)
> http://localhost:4200/#state=eyJ…
> ```
> 
> When it doesn't work:
> 
> ```
> about:blank (several times)
> http://localhost:4200/# (without anything after the #)
> ```
> 
> The problem might be with the POLLING_INTERVAL_MS. If I increase it from the default 50ms to 500ms I'm not able to sign in whatsoever. If I decrease it to 5ms, it seems to be working every time.
> 
> So it seems that something is clearing the query string from the URL before the library extracts it (the query string is available for less than 50ms), and it's based on a pure chance whether it works or not.

Can you share how this was changed?

**Comment by jasonnutter** [comment]
> I logged the value of the **href** variable in the **monitorIframeForHash** function.
> 
> When it works:
> 
> ```
> about:blank (several times)
> http://localhost:4200/#state=eyJ…
> ```
> 
> When it doesn't work:
> 
> ```
> about:blank (several times)
> http://localhost:4200/# (without anything after the #)
> ```
> 
> The problem might be with the POLLING_INTERVAL_MS. If I increase it from the default 50ms to 500ms I'm not able to sign in whatsoever. If I decrease it to 5ms, it seems to be working every time.
> 
> So it seems that something is clearing the query string from the URL before the library extracts it (the query string is available for less than 50ms), and it's based on a pure chance whether it works or not.

This would indicate that something is changing/removing the hash before the library has a chance to parse it. Are you using hash routing in your Angular application?

**Comment by ZacharyHiggins-dbMotion** [comment]
> > I logged the value of the **href** variable in the **monitorIframeForHash** function.
> > When it works:
> > ```
> > about:blank (several times)
> > http://localhost:4200/#state=eyJ…
> > ```
> > 
> > 
> > When it doesn't work:
> > ```
> > about:blank (several times)
> > http://localhost:4200/# (without anything after the #)
> > ```
> > 
> > 
> > The problem might be with the POLLING_INTERVAL_MS. If I increase it from the default 50ms to 500ms I'm not able to sign in whatsoever. If I decrease it to 5ms, it seems to be working every time.
> > So it seems that something is clearing the query string from the URL before the library extracts it (the query string is available for less than 50ms), and it's based on a pure chance whether it works or not.
> 
> @anth-git This would indicate that something is changing/removing the hash before the library has a chance to parse it. Are you using hash routing in your Angular application?

Regardless of @anth-git 's answer. We are having the same issue and using hash routing. It's required for our angular project to work correctly with Azure Static Web Apps Preview (justifying our reasoning).

**Comment by anth-git** [comment]
@ZacharyHiggins-dbMotion I changed it in node_modules\msal\lib-es6\utils\WindowUtils.js (obviously just for investigation purposes)

@jasonnutter Yes, I'm using hash location strategy in the Angular app.

**Comment by anth-git** [comment]
I turned off the hash routing and the behavior is the same, so I don't think it's related to the routing strategy.

**Comment by anth-git** [comment]
I think I managed to get it working even for the high POLLING_INTERVAL_MS values:
```
@NgModule({
  imports: [RouterModule.forRoot(routes,
    {
      enableTracing: false,
      useHash: true,
      initialNavigation: isInIframe() ? 'disabled' : undefined // <- THIS
    })],
  exports: [RouterModule]
})
export class AppRoutingModule {
}

export function isInIframe() {
  return window !== window.parent && !window.opener;
}
```

**Comment by abellisai** [comment]
Hi @anth-git 
Thanks for your suggestion testing on my personal env, it works!

**Comment by ZacharyHiggins-dbMotion** [comment]
> I think I managed to get it working even for high POLLING_INTERVAL_MS values:
> 
> ```
> @NgModule({
>   imports: [RouterModule.forRoot(routes,
>     {
>       enableTracing: false,
>       useHash: true,
>       initialNavigation: isInIframe() ? 'disabled' : undefined // <-THIS
>     })],
>   exports: [RouterModule]
> })
> export class AppRoutingModule {
> }
> 
> export function isInIframe() {
>   return window !== window.parent && !window.opener;
> }
> ```

https://angular.io/api/router/InitialNavigation

This makes sense. Nice find. I'm testing this locally as well. So far so good. Will follow up tomorrow (unless the problem returns before then) with an update. Thanks!

**Comment by jasonnutter** [comment]
Yes, preventing navigation when you are in an iframe is a way to mitigate the behavior from the router that removes the hash. We show a similar workaround in the samples, but we'll make sure this is better documented.

**Comment by piotrduperas** [comment]
Hello @jasonnutter !
We are experiencing the same problem. This is one of our logs from console:
`Wed, 29 Jul 2020 09:42:43 GMT:1234-1.3.3-Verbose-pii RenewToken scope and authority: https://<our.scope>|undefined`
I have investigated this `undefined`, it is a `request.authority` field, but I was unable to trace it further.

We have found that this can be caused by MsalGuard on redirect page, but removing it and calling `msalService.loginRedirect()` does not work. The suggestion by @anth-git does not work for us as well.

**Comment by anth-git** [comment]
@piotrduperas Strange it doesn't work. It's pretty foolproof for me now.
Do you use approach described in https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-js-avoid-page-reloads ?
Also make sure you don't call any msal methods (like loginRedirect, acquireTokenSilent) in the iframe.

**Comment by piotrduperas** [comment]
@anth-git We are observing that this approach works on Firefox, but on Chrome and Edge it works only sometimes, however, checking "Disable cache" in Dev Tools results in much rarer occurrences of this error. We tried to use some interceptor that sets browser flags concerned about cache, but it did not work as well.

**Comment by alvipeo** [comment]
Does anybody know if this issue exists in 2.0?

**Comment by alvipeo** [comment]
@anth-git I just added MsalComponent and it still doesn't work with my live.com account :/

**Comment by arnevdv** [comment]
> @ashishbhulani You should be able to handle this without modifying the interceptor. When you make the http request and you receive an error, if it is a timeout or interaction required error, invoke acquireTokenPopup with the same set of scopes, and then make the request again.
> 
> ```js
> const GRAPH_ENDPOINT = "https://graph.microsoft.com/v1.0/me";
> 
> this.http.get(GRAPH_ENDPOINT)
>     .subscribe({
>       next: (profile) => {
>         this.profile = profile;
>       },
>       error: (err: AuthError) => {
>           this.authService.acquireTokenPopup({
>             scopes: this.authService.getScopesForEndpoint(GRAPH_ENDPOINT)
>           })
>           .then(() => {
>             this.http.get(GRAPH_ENDPOINT)
>               .toPromise()
>               .then(profile => {
>                 this.profile = profile;
>               });
>           });
>       }
>     });
> ```
> 
> Note, if you want to use `acquireTokenRedirect` or `loginRedirect` instead, your application will need to implement `handleRedirectCallback` separately, not inside the interceptor or where you make the http request. Instead, it needs to be invoked on page load, [as demonstrated in the Angular 8 sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/angular8-sample-app/src/app/app.component.ts#L27). Remember, if you call redirect, the browser will fully redirect away from your application and lose all context, which is why we recommend acquireTokenPopup instead for this scenario.

Does anyone has working link for the angular 8 example?

**Comment by alvipeo** [comment]
I ended up changing to another library and it works like a charm.

**Comment by lukan95** [comment]
I try @anth-git approach. If i run it on localhost, it works fine. But if i try run it on a production server, only a blank page will be displayed. 

Im using also hash strategy and  msalguards.

 Is there a need for a refresh or something  else?

**Comment by ZacharyHiggins-dbMotion** [comment]
> I try @anth-git approach. If i run it on localhost, it works fine. But if i try run it on a production server, only a blank page will be displayed.
> 
> Im using also hash strategy and msalguards.
> 
> Is there a need for a refresh or something else?

This was our exact scenario as well. Hash routing, works locally, does not work in Azure Static Web. I tacked it again last night and was able to resolve/workaround our issues. Some things to note that I was doing that differed from the example code:

- Immediately on page load, I was checking to see if the user was authenticated, and perfoming a **this.authService.loginRedirect();** if not. 
- Once the user was authenticated, I redirected the user to a page which was set to activate MSALGuard.
```
    if(this.MSALGetAccountState()){
      this.router.navigateByUrl('main')
    } else {
      this.MSALlogin();
    }
```
- I tried handling the redirect from our AD app config in Azure (wrong according to the logs), and within the app component. Both yielded the same results.
- One of our project requirements was to force a login, every time the site was accessed, so wrapping this in the appcomponent page seems like a good way to go. Today, it appears this was a bad idea.

Using the example here (without deviation) https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-angular-samples/angular9-sample-app, I was able to resolve this error, albeit deviating from our original project spec.

Looking at my solution, compared to others, this error appears to throw on wide variety of problems, likely due to a knowledge issue with how MSAL works (there appear to be a lot of gotcha's with this library). If anyone is interested, I'm willing to share our previous source code that produced this error, but tbh, I probably messed something up due to a knowledge gap.

**Comment by anth-git** [comment]
@ZacharyHiggins-dbMotion 
I've learned the hard way that a most important thing when dealing with msal is to not make any redirection or authentication calls when in an iframe. It can be easily checked by comparing window !== window.parent.

Apart from that, why did you have to redirect manually? It should be redirected to redirectUri address (the url in the Azure must match with msal config) Are you sure you had it correctly set in a production?

To force login at every page load, you need to pass { prompt: 'select_account' } to authentication methods (e.g. loginRedirect({ prompt: 'select_account' }))

**Comment by ZacharyHiggins-dbMotion** [comment]
> @ZacharyHiggins-dbMotion
> I've learned the hard way that the most important thing when dealing with msal is to not make any redirection or authentication calls when in an iframe. It can be easily checked by comparing window !== window.parent.
> 
> Apart from that, why did you have to redirect manually? It should be redirected to redirectUri address (the url in the Azure must match with msal config) Are you sure you had it correctly set in a production?
> 
> To force login at every page load, you need to pass { prompt: 'select_account' } to authentication methods (e.g. loginRedirect({ prompt: 'select_account' }))

Hi @anth-git , thanks for the follow up. Regarding the redirect manually question. I had my redirectURI set to a page that activated MSALGuard, but received a *warning* that said this was a bad idea. I'm positive I had the redirectURI's matching in prod, mainly because that issue would throw a more specific error at me when I forgot to change it in the config or Azure, and it's still working. It's worth noting that I did change it back to /#/main in our working tree, and I did not get the warning... so this was likely also me misinterpreting the problem due to our deviations coupled with the fact this is our first web project (attempting to modernize an application).

Thanks for the additional info with login redirect. I will give that a shot. If I have any issues, I'll do some research and open a separate issue if needed. Poor @jasonnutter is probably going crazy with this one.

**Comment by anth-git** [comment]
I'm not sure if understand what does it mean that you had _redirectURI set to a page that activated MSALGuard_ .

In the dev you should just have it set to http://localhost:4200. In the prod it should be set to the prod address. 

Any redirections after the authentication should be handled using Angular routing:
```{ path: '', redirectTo: 'main', pathMatch: 'full', canActivate: [MsalGuard] },```

In a simple scenario you don't even need to invoke any Msal methods manually, but just let MsalGuard to handle the authentication (in such case you also need to pass { prompt: 'select_account' } to MsalAngularConfiguration.extraQueryParameters config, if you want force login on every page load)

**Comment by ZacharyHiggins-dbMotion** [comment]
> @ZacharyHiggins-dbMotion
> I'm not sure if understand what does it mean that you had _redirectURI set to a page that activated MSALGuard_ .
> 
> In the dev you should just have it set to http://localhost:4200. In the prod it should be set to the prod address.
> 
> Any redirections after the authentication should be handled using Angular routing:
> `{ path: '', redirectTo: 'main', pathMatch: 'full', canActivate: [MsalGuard] },`
> 
> In a simple scenario you don't even need to invoke any Msal methods manually, but just let MsalGuard to handle the authentication (in such case you also need to pass { prompt: 'select_account' } to MsalAngularConfiguration.extraQueryParameters config, if you want force login on every page load)

That's correct. To be clear, my issue **was not** with configuring a matching redirectURI in Azure, my appmodule.ts (MSALConfigFactory), or the intricacies of managing different addresses between a prod, local dev environment, and the others.

The redirectURI I'm referring to is the one configured in our MSALConfigFactory (and obviously matching in our Azure AD app config)

```
function MSALConfigFactory(): Configuration {
  return {
    auth: {
      clientId: environment.ClientId,
      authority: environment.AuthorityServer,
      validateAuthority: true,
      redirectUri: environment.RedirectURL,
      postLogoutRedirectUri: "",
      navigateToLoginRequestUrl: false,
    },
    cache: {
      cacheLocation: "sessionStorage",
      storeAuthStateInCookie: false, // set to true for IE 11
    },
    system: {
      loadFrameTimeout: 30000,
    }
  };
}
```

This was a separate issue though, and it appears to only have surfaced because of our previous implementation. Since the timeout error wasn't very intuitive output for our particular issue, we had to do some discovery and make some assumptions (and a lot of reading). The way achieve our original spec is pretty clear now given your advice (thank you again).

To reiterate, our original implementation was incorrect because we were trying to trigger the login, and handle the routing programmatically in the appcomponent.ts, None of this was needed and we just had to set the app router up correctly.

If you look at the demo project... there is a "login" button, and some code behind it. I was essentially running that in ngOnInit. Obviously not right, but was not immediately clear until I had a chance to sit down and look at it.

**Comment by ZacharyHiggins-dbMotion** [comment]
Clarification - We wanted the user to be automatically presented a login page, which is why we essentially added loginredirect() in ngOnInit in appcomponent. Regarding the passing the select_account param in the loginredirect method. I think this was part of our problem (it's been a while), and why we were doing this in ngOnInit... so we could pass these parameters. Our request actually looked like this:

```
  request = {
    scopes: ["user.read"],
    prompt: 'login'
  }
```

It's still not clear how with our current implementation, that we can force this if we aren't actually calling loginRedirect({ prompt: 'select_account' })) anywhere.... I will need to do more research. It's likely very simply, but that particular workflow is no longer a priority for this project, but this seems out of scope with the issue being tracked here. Our conversation definitely helped with clarifying the proper implementation.

**Comment by anth-git** [comment]
@ZacharyHiggins-dbMotion As for _select_account_, add it to MSAL_CONFIG_ANGULAR provider factory:
```
{
        provide: MSAL_CONFIG_ANGULAR,
        useFactory: MSALAngularConfigFactory
},
unction MSALAngularConfigFactory(): MsalAngularConfiguration {
  return {      
      extraQueryParameters: {
        prompt: 'select_account'
      }
    }
}
```

**Comment by ZacharyHiggins-dbMotion** [comment]
> @ZacharyHiggins-dbMotion As for _select_account_, add it to MSAL_CONFIG_ANGULAR provider factory:
> 
> ```
> {
>     provide: MSAL_CONFIG_ANGULAR,
>     useFactory: MSALAngularConfigFactory
> }
> 
> function MSALAngularConfigFactory(): MsalAngularConfiguration {
>   return {      
>       extraQueryParameters: {
>         prompt: 'select_account'
>       }
>     }
> }
> ```

My friend, you have been very very helpful today. I haven't even started Googling that particular problem yet. Thank you very very much. I'll give that a shot in about an hour :)

**Comment by ZacharyHiggins-dbMotion** [comment]
> @ZacharyHiggins-dbMotion As for _select_account_, add it to MSAL_CONFIG_ANGULAR provider factory:
> 
> ```
> {
>     provide: MSAL_CONFIG_ANGULAR,
>     useFactory: MSALAngularConfigFactory
> }
> 
> function MSALAngularConfigFactory(): MsalAngularConfiguration {
>   return {      
>       extraQueryParameters: {
>         prompt: 'select_account'
>       }
>     }
> }
> ```

Following up. Everything is running flawlessly now. Thank you very much! Have a great weekend! 👍

**Comment by gunnamvasavi** [comment]
Hi @ZacharyHiggins-dbMotion and @anth-git.... I'm also facing the same issue and the scenario is like users have to be logged in for all the page routes and using MSALGUARD for that. Could you explain me if any workaround is there to avoid "token renewal failed due to timeout". And my code is as follows:

app.module.ts:
` MsalModule.forRoot({
      auth: {
  
        clientId: environment.clientId,
        authority: "https://login.microsoftonline.com/"+environment.tenatId,
        validateAuthority: true,
        redirectUri: environment.REDIRECT_URI,
        postLogoutRedirectUri: environment.REDIRECT_URI,
        navigateToLoginRequestUrl: true,
      },
      cache: {
        cacheLocation: 'localStorage',
        storeAuthStateInCookie: isIE, // set to true for IE 11
      },
    },
      {
        popUp: false,
        consentScopes: [
          'user.read',
          'openid',
          'profile',
        ],
        unprotectedResources: [],
        protectedResourceMap: [
          ['https://graph.microsoft.com/v1.0/me', ['user.read']]
        ],
        extraQueryParameters: {}
      })
  ],
  providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: TokenInterceptor,
      multi: true
    },
    { provide: HTTP_INTERCEPTORS, useClass: MsalInterceptor, multi: true },
  ],`


routing-module:

`const routes: Routes = [
  // { path: '', redirectTo: '/EL', pathMatch: 'full' },
  {
    path: '',
    component: ExecHomeComponent,
    canActivate: [MsalGuard],
    children:[
      {
        path: ':main/:sub',
        component: OtherPageComponent,
        canActivate: [MsalGuard],
      },
      {
        path: ':main/:sub/:classification',
        component: OtherPageComponent,
        canActivate: [MsalGuard],
      },
      {
        path:'',
        component:LandingComponent,
        canActivate: [MsalGuard],
      },  
    ]
  }
];`

based on the role i'm handling some redirects in app.component.ts ngOnInit
`  if (!role) {
      let res = await this.authentication.getRole();
      role=localStorage.getItem('role');
      if(role){
        this.roleReady = true;
        this.router.navigate([this.globalListen.navList[role]['link']]);
      }
    } else {
      this.roleReady = true;
      if(this.path=='/')
      this.router.navigate([this.globalListen.navList[role]['link']]);
    }`

Help me out! As we are launching this to a bunch of users in the next I wanted to fix this issue.

**Comment by anth-git** [comment]
@gunnamvasavi
Don't navigate when in an iframe:

```
if (isInIframe()){
   return;
}

 if (!role) {
      let res = await this.authentication.getRole();
      role=localStorage.getItem('role');
      if(role){
        this.roleReady = true;
        this.router.navigate([this.globalListen.navList[role]['link']]);
      }
    } else {
      this.roleReady = true;
      if(this.path=='/')
      this.router.navigate([this.globalListen.navList[role]['link']]);
    }

function isInIframe() {
  return window !== window.parent && !window.opener;
}
```

Alternatively you can implement this https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-js-avoid-page-reloads

Apart from that you can disable initialNavigation: https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/1592#issuecomment-664818059

**Comment by lukan95** [comment]
> I try @anth-git approach. If i run it on localhost, it works fine. But if i try run it on a production server, only a blank page will be displayed.
> 
> Im using also hash strategy and msalguards.
> 
> Is there a need for a refresh or something else?

@anth-git  I found a difference between the local (dev) version and the production version. In the dev version of the build is aot set to false. At the production I have aot turned on. If I set in production build aot to false, it works for me, , but otherwise routing probably doesn't work.

**Comment by anth-git** [comment]
@lukan95 It works for me with AOT just fine, besides this is a default setting since Angular 9. 
I don't know why it could make a difference in you case though. 

Assuming that your issue is related to a timeout error, did you try to change POLLING_INTERVAL_MS to some low value, for example 5ms (node_modules\msal\lib-es6\utils\WindowUtils.js). This way you could find out if something removes hash before library can parse it or is it something else.

**Comment by lukan95** [comment]
@anth-git I am using Anguler 8 and there is default setting without aot. So I set it in tsconfig.json. Maybe I'll try to update angular to version 9.

Yes i tried and it was okay.

**Comment by gunnamvasavi** [comment]
@anth-git your approach worked for me. Tried in local and prod as well its working! Thanks for sharing work around.

**Comment by lukan95** [comment]
I changed the angular version from 8 to 10 and the problem with active aot  is no longer and the @anth-git approach works flawlessly. Thank you.

**Comment by Rookian** [comment]
> @ashishbhulani You should be able to handle this without modifying the interceptor. When you make the http request and you receive an error, if it is a timeout or interaction required error, invoke acquireTokenPopup with the same set of scopes, and then make the request again.
> 
> ```js
> const GRAPH_ENDPOINT = "https://graph.microsoft.com/v1.0/me";
> 
> this.http.get(GRAPH_ENDPOINT)
>     .subscribe({
>       next: (profile) => {
>         this.profile = profile;
>       },
>       error: (err: AuthError) => {
>           this.authService.acquireTokenPopup({
>             scopes: this.authService.getScopesForEndpoint(GRAPH_ENDPOINT)
>           })
>           .then(() => {
>             this.http.get(GRAPH_ENDPOINT)
>               .toPromise()
>               .then(profile => {
>                 this.profile = profile;
>               });
>           });
>       }
>     });
> ```
> 
> Note, if you want to use `acquireTokenRedirect` or `loginRedirect` instead, your application will need to implement `handleRedirectCallback` separately, not inside the interceptor or where you make the http request. Instead, it needs to be invoked on page load, [as demonstrated in the Angular 8 sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/angular8-sample-app/src/app/app.component.ts#L27). Remember, if you call redirect, the browser will fully redirect away from your application and lose all context, which is why we recommend acquireTokenPopup instead for this scenario.

The link does not work anymore. Is there a complete working example for angular anywhere?

**Comment by jasonnutter** [comment]
Updated link: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-angular-samples/angular8-sample-app/src/app/app.component.ts

**Comment by vamshinarra** [comment]
This issue is still present no change after using the iframe condition on router outlet

**Comment by pawlikowski-michal** [comment]
Does anyone was able to solve this issue using Angular 8? We tried almost every solution from this thread, but with no luck...

**Comment by jwahr** [comment]
☝️ same. I get the error once in awhile, not all the time. Not good UX.

**Comment by gunnamvasavi** [comment]
On the routing @anth-git approach works fine, not seeing any issues but from the Interceptor side its not working. For Example I'm in a particular page and the token expires, clicking on some button in the UI would do an API call and at this point of time:

      Expected behavior: MsalInterceptor should take care of the token expiry and would give refreshed token.
      Actual behavior: Its throwing the same client token renewal failed error.

 @jasonnutter and @anth-git are there any workarounds for this behavior?

Note: The latest version of MSAL@1.4.0 also not working for me. Its giving token renewal failed for each page routing and in Incognito mode it went on infinite loop.

**Comment by joshpitkin** [comment]
Ok, I spent a day on this so I will share some findings and in case it helps anyone else..

In my implementation I don't need/want any special consent scopes yet, just openid and profile which I believe are default.

However the loginRedirect() -> [saveAccessToken()](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/70b87bcd27bfe67789c92b9bc046b45733f490ed/lib/msal-core/src/UserAgentApplication.ts#L1652) method doesn't ever save those values in the "scope" property of the JSON object key for the tokens, because the value is undefined and it the key is JSON.stringified.  

Then whenever a call to the acquireTokenSilent() method is called it doesn't find any cached tokens!
Tracked this down to [getCachedToken()](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/70b87bcd27bfe67789c92b9bc046b45733f490ed/lib/msal-core/src/UserAgentApplication.ts#L1363) -> [getAllAccessTokens()](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/70b87bcd27bfe67789c92b9bc046b45733f490ed/lib/msal-core/src/cache/AuthCache.ts#L189) which has a condition looking for the "scopes" property to exist in the key from the storage items.  Later it also looks for the value to match the requested scope.

The result is that every time I call acquireTokenSilent() it queues up another login redirect request in the hidden iframe, and eventually too many back-to-back repeat calls were causing intermittent "timeouts" from MSAL in my case.

So is this intentional / by design or do I have something setup wrong?  Additional consent scopes are not required are they?

As a workaround I am duplicating that storage item and adding a "scopes" property to it with the values of "openid" and "profile", but I have to re-create that every time the tokens are not loaded from cache.

**Comment by tnorling** [comment]
@joshpitkin `loginRedirect` only requests an idToken, not an accessToken. The cache entry you are seeing that has scopes missing is actually the idToken. So when you call `acquireTokenSilent` it's looking for accessTokens which, by design, do not exist yet. We will be addressing this in #2206, which will update `acquireTokenSilent` to look for and return an idToken if that is what was requested. 

If your use case requires an accessToken you should call an interactive `acquireToken` method first, i.e. `acquireTokenPopup` or `acquireTokenRedirect`. If you are using `msal@1.4.0` and you require only idTokens then this is a bug that will be addressed in the above PR. The workaround for now would be to downgrade to 1.3.4 in addition to calling the interactive methods first.

**Comment by joshpitkin** [comment]
thank you @tnorling this helps explain a lot.  for now I will stick with 1.3.4 until we can use `acquireTokenSilent` to refresh just the idToken.

**Comment by rangeranger** [comment]
@anth-git I'm not able to tag you in my issue directly. So, commenting here. Can you please look into [this ](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/2234#issue-694570747)issue? I configured exactly how you mentioned. I still get some errors.

**Comment by veera-raghavagupta** [comment]
> Hi @ZacharyHiggins-dbMotion and @anth-git.... I'm also facing the same issue and the scenario is like users have to be logged in for all the page routes and using MSALGUARD for that. Could you explain me if any workaround is there to avoid "token renewal failed due to timeout". And my code is as follows:
> 
> app.module.ts:
> 
> ```
> 
>  MsalModule.forRoot({
>       auth: {
>         clientId: environment.clientId,
>         authority: "https://login.microsoftonline.com/"+environment.tenatId,
>         validateAuthority: true,
>         redirectUri: environment.REDIRECT_URI,
>         postLogoutRedirectUri: environment.REDIRECT_URI,
>         navigateToLoginRequestUrl: true,
>       },
>       cache: {
>         cacheLocation: 'localStorage',
>         storeAuthStateInCookie: isIE, // set to true for IE 11
>       },
>     },
>       {
>         popUp: false,
>         consentScopes: [
>           'user.read',
>           'openid',
>           'profile',
>         ],
>         unprotectedResources: [],
>         protectedResourceMap: [
>           ['https://graph.microsoft.com/v1.0/me', ['user.read']]
>         ],
>         extraQueryParameters: {}
>       })
>  providers: [
>     {
>       provide: HTTP_INTERCEPTORS,
>       useClass: TokenInterceptor,
>       multi: true
>     },
>     { provide: HTTP_INTERCEPTORS, useClass: MsalInterceptor, multi: true },
>   ],
> ```
> 
> routing-module:
> 
> ```
> const routes: Routes = [
>   // { path: '', redirectTo: '/EL', pathMatch: 'full' },
>   {
>     path: '',
>     component: ExecHomeComponent,
>     canActivate: [MsalGuard],
>     children:[
>       {
>         path: ':main/:sub',
>         component: OtherPageComponent,
>         canActivate: [MsalGuard],
>       },
>       {
>         path: ':main/:sub/:classification',
>         component: OtherPageComponent,
>         canActivate: [MsalGuard],
>       },
>       {
>         path:'',
>         component:LandingComponent,
>         canActivate: [MsalGuard],
>       },  
>     ]
>   }
> ];
> ```
> 
> based on the role i'm handling some redirects in app.component.ts ngOnInit
> 
> ```
> 
>  if (!role) {
>       let res = await this.authentication.getRole();
>       role=localStorage.getItem('role');
>       if(role){
>         this.roleReady = true;
>         this.router.navigate([this.globalListen.navList[role]['link']]);
>       }
>     } else {
>       this.roleReady = true;
>       if(this.path=='/')
>       this.router.navigate([this.globalListen.navList[role]['link']]);
>     }
> ```
> 
> Help me out! As we are launching this to a bunch of users in the next week, I wanted to fix this issue.

Can you please help if you have fixed the issue as I am facing same issue. I am using msal 1.4 and @azure/msal-angular 1.1.0 and custom interceptor

**Comment by vishmay-AF** [comment]
My question here is why does MSAL guard need to renew the token on each page request? even if the old token is not yet expired?

**Comment by jmckennon** [comment]
@vishmay-AF This should have been fixed in `msal@1.4.1`, can you upgrade to the new version and if your problem still persists, open a new issue with your version, config, and usage?

**Comment by vishmay-AF** [comment]
@jmckennon 
Just now, my issues was solved by setting up azure configuration, 
I'm not sure which one did the trick, so I'm listing all... 
1. I gave some permission, (attached)
2. created client secret, 
3. Exposed API with scope for impersonate user.
I'ven't checked with 1.4.1, 
1.4.0 is working for me.
[permissions .txt](https://github.com/AzureAD/microsoft-authentication-library-for-js/files/5334917/new.5.txt)

**Comment by rpapeters** [comment]
Also getting token timeout error now and then. v1.4.0 was worse and v1.4.1 is better but also has other issues which should be resolved in 1.4.2 but not yet released. Could the origin of this problem be in the fact that both the MsalGuard and MsalInterceptor are trying to get a token very shortly after each other? (one for canActivate and one for a back-end request)

**Comment by vishmay-AF** [comment]
The occrance has been reduced since 1.4.1 but I'm still getting errors in some cases. 
it happens on the rediredction page after login. not all the time, it occurs for few times. 
mostly 1 out of 3 tries fails due to this and it affects my other working code too.
this time MSAL gaurd does not cause trouble i think interceptor is the reason.

let me know if you have any solutions

**Comment by rpapeters** [comment]
I've been able to resolve the token timeout issue by following the common issues guide found here: 
[Common Issues](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-core/docs/FAQ.md#common-issues)

In our setup we had a redirect in the angular router from '/' to '/some-url' and we where redirecting msal to '/', which would trigger a redirect by angular while acquiring the token. Msal did not like that. Now we have a redirect to (in our case) '/msal' for msal, without angular redirecting under the hood and the problem seems to be resolved. Don't forget to add the '/msal' as an authenticated Redirect URI in the app registration in Azure Ad (B2C).

**Comment by derekparsons718** [comment]
I hoped I wouldn't have to, but I got so tired of this bug that I ended up just upgrading to [v2 (msal-browser)](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-browser) about a week ago, and I haven't seen the issue since.

**Comment by rangeranger** [comment]
@derekparsons718 : Did you write a custom interceptor to add tokens to HTTP requests?

**Comment by mattmesser** [comment]
@jasonnutter @tnorling think this issue/library is going to be fixed before the end of the year? It's absolutely bricking our spa. Our team is starting to catch a lot of heat for it since we put it in front of testers and they cant log in over 60% of the time. That, and issue #2492 have made testing our SPA impossible. It's a nightmare.

**Comment by jasonnutter** [comment]
> @jasonnutter @tnorling think this issue/library is going to be fixed before the end of the year? It's absolutely bricking our spa. Our team is starting to catch a lot of heat for it since we put it in front of testers and they cant log in over 60% of the time. That, and issue #2492 have made testing our SPA impossible. It's a nightmare.

We are looking to merge #2492 soon (hopefully today). 

Have you done any debugging into what is causing the timeouts (they can happen for many reasons)? A common problem with Angular applications is that the Angular router clears the hash before MSAL has a chance to parse it. Setting `initialNavigation` to false when your app is inside an iframe has shown to help mitigate that specific root cause.

**Comment by jasonnutter** [comment]
If that does not help, feel free to share MSAL logs and fiddler traces with me via email (which is in my profile).

**Comment by daviddejaeger** [comment]
I'm also still getting this error. Refreshing one or several times solves the problem but that is not really an option. I'm not doing anything special and following the demo samples. One difference is that the login code is inside a base.component because I have a resolver associated with that component with an api call to my backend for initialization logic instead of app.component.

app.component.ts:
`@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit {
  constructor() { }

   ngOnInit() {

   }
}`
app.component.html:
`<router-outlet></router-outlet>`

app.module.ts:
`const routes: Routes = [
  { path: '', 
    component: BaseComponent, 
    canActivate: [MsalGuard],
    children: [ 
      { path: '', component: DeviationlistComponent, canActivate: [MsalGuard], },
      { path: 'unauthorized', component: UnauthorizedComponent, canActivate: [MsalGuard], },
      { path: 'myAssignments', component: MyassignmentsComponent, canActivate: [MsalGuard], },
      { path: 'deviations', component: DeviationlistComponent, canActivate: [MsalGuard], },
      { path: 'deviations/:id', component: DeviationdetailsComponent, canActivate: [MsalGuard], },
      { path: 'success', component: SuccessComponent, canActivate: [MsalGuard], },
      { path: 'CreateDeviation', component: DeviationcreateComponent, canActivate: [MsalGuard], },
      { path: 'deviationsfollowup', component: DeviationsFollowupComponent, canActivate: [MsalGuard,RoleGuardService,], data: { expectedRoles: ['Admin','QualityMember','QualityManager'] }},
      { path: 'charts', component: ChartsComponent, canActivate: [MsalGuard,RoleGuardService,], data: { expectedRoles: ['Admin','QualityMember','QualityManager'] }}
    ],
    resolve: {
      data: AuthenticatedAppResolver
    }
  },`

base.component.ts:
`export class BaseComponent implements OnInit {
  subscriptions: Subscription[] = [];

  title = 'Deviation Management Tool';
  username = '';
  isIframe = false;
  loggedIn = false;

  constructor(private broadcastService: BroadcastService, private authService: MsalService, private userService: UserService) { 
    locale(navigator.language);
  }

  ngOnInit() {
    let loginSuccessSubscription: Subscription;
    let loginFailureSubscription: Subscription;

    this.isIframe = window !== window.parent && !window.opener;

    this.checkAccount();

    loginSuccessSubscription = this.broadcastService.subscribe('msal:loginSuccess', () => {
      this.checkAccount();     
    });

    loginFailureSubscription = this.broadcastService.subscribe('msal:loginFailure', (error) => {
      console.log('Login Fails:', error);
    });

    this.broadcastService.subscribe('msal:logOut', () => {
      this.userService.removeUser();
    });

    this.subscriptions.push(loginSuccessSubscription);
    this.subscriptions.push(loginFailureSubscription);

    this.authService.handleRedirectCallback((authError, response) => {
      if (authError) {
        console.error('Redirect Error: ', authError.errorMessage);
        return;
      }

      console.log('Redirect Success: ', response.accessToken);
    });

    this.authService.setLogger(new Logger((logLevel, message, piiEnabled) => {
      console.log('MSAL Logging: ', message);
    }, {
      correlationId: CryptoUtils.createNewGuid(),
      piiLoggingEnabled: false
    }));
  }

  ngOnDestroy(): void {
    this.subscriptions.forEach((subscription) => subscription.unsubscribe());
  }

  checkAccount() {
    this.loggedIn = !!this.authService.getAccount();
  }

  login() {
    const isIE = window.navigator.userAgent.indexOf('MSIE ') > -1 || window.navigator.userAgent.indexOf('Trident/') > -1;

    if (isIE) {
      this.authService.loginRedirect();
    } else {
      this.authService.loginPopup();
    }
  }

  logout() {
    this.authService.logout();
  }`

base.component.html:
`<div class="sidebar">
    <app-navmenu></app-navmenu>
  </div>
  
  <div class="main">
    <div class="top-row px-4 mb-4">
      <app-companyselector></app-companyselector>
      Hello, {{username}}
      <a [routerLink]="" *ngIf="!loggedIn" (click)="login()">Login</a>
      <a [routerLink]="" *ngIf="loggedIn" (click)="logout()">Logout</a>
    </div>   
    <div class="content px-4">
      
      <router-outlet *ngIf="!isIframe"></router-outlet>
    </div>
  </div>
  `

role-guard.service.ts:
`export class RoleGuardService implements CanActivate {

  constructor(private authService: MsalService, private router: Router) {}
  
  canActivate(route: ActivatedRouteSnapshot): boolean {
    const expectedRoles = route.data.expectedRoles;
    let authorized = false;

    expectedRoles.forEach(element => {
      if (this.authService.getAccount().idTokenClaims.roles.includes(element)) {      
        authorized = true;
      }
    });

    if (authorized){
      return true;
    }
    else{
      if (!this.isInIframe()){
        return false;
      }
      else{
        this.router.navigate(['unauthorized']);
        return false;
      }     
    }   
  }

  isInIframe() {
    return window !== window.parent && !window.opener;
  }
}`

authenticatedAppResolver.ts:
`@Injectable()
export class AuthenticatedAppResolver implements Resolve<Observable<User>> {

    public constructor(private userService: UserService) { }

    public resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<User> {
        return this.userService.initializeUser();
    }
}`

I wrapped all my API calls with the suggestion 'When you make the http request and you receive an error, if it is a timeout or interaction required error, invoke acquireTokenPopup with the same set of scopes, and then make the request again.' For example:
`getCategories(): void {
    this.categoryService.getCategories().subscribe({next: (categories) => {
      this.categories = categories 
      this.changesMade = false;    
    },
    error: (err: AuthError) => {
      this.authService.acquireTokenPopup({
        scopes: this.authService.getScopesForEndpoint(environment.resources.deviationApi.resourceUri)
      })
      .then(() => {
        this.categoryService.getCategories()
          .toPromise()
          .then(categories => {
            this.categories = categories
            this.changesMade = false;
          });
      });
    }
  });
  }`

However for the data-resolver I can't manage to do that because it requires an Observable. I'm a bit lost on how to proceed. Am I making an obvious mistake somewhere? Any help is appreciated.

**Comment by mattmesser** [comment]
> > @jasonnutter @tnorling think this issue/library is going to be fixed before the end of the year? It's absolutely bricking our spa. Our team is starting to catch a lot of heat for it since we put it in front of testers and they cant log in over 60% of the time. That, and issue #2492 have made testing our SPA impossible. It's a nightmare.
> 
> We are looking to merge #2492 soon (hopefully today).
> 
> Have you done any debugging into what is causing the timeouts (they can happen for many reasons)? A common problem with Angular applications is that the Angular router clears the hash before MSAL has a chance to parse it. Setting `initialNavigation` to `false` or `disabled` when your app is inside an iframe has shown to help mitigate that specific root cause. Example:
> 
> https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/9fb15d5cc5ed4203f21ac067ea847d4592fd895c/samples/msal-angular-v2-samples/angular10-browser-sample/src/app/app-routing.module.ts#L32

@jasonnutter  Thanks for the 'initialNavigation' suggestion.  I'm hesitant to say it's definitively fixed yet: it usually occurs only after token expiration, which takes 24hrs. But things have improved significantly.

Just a note: In the config I also changed cacheLocation to 'sessionStorage' from 'localStorage'. I think this helped ameliorate several other issues that were cropping up.

To answer your question, yes I have spent quite a lot of time cumulatively debugging this. To me, it's been exceptionally difficult to precisely determine root cause. I've been unable to consistently replicate the error despite my best efforts.  The timeout error bricks my dev-tools so my tried and true "debugger;" spamming did not work.

I'll reach out if things catch on fire again, or post here to confirm that the fix worked by Friday.

Thanks again!

**Comment by jasonnutter** [comment]
Summarizing recommended mitigations for v1:

- Set `initialNavigation` in `app-routing.module.ts` to `'disabled'` when your app is loaded in an iframe
- Increase the timeout set in `loadFrameTimeout`
- Use a blank page as a redirect URI
- Use the latest version of msal (1.4.4)

The best mitigation is to upgrade to MSAL Angular v2, as we know have the first alpha version available: https://github.com/AzureAD/microsoft-authentication-library-for-js/releases/tag/msal-angular-v2.0.0-alpha.0

I'm going to close this issue. If your Angular app continues to experience timeouts after the above mitigations, please open a new issue, thanks!

**Comment by skironDotNet** [comment]
> @jasonnutter I'm getting the same error. msal 1.3.1., msal-angular 1.0.0.
> 
> Auth only works right after I clear the browser cache. Hours later:
> 
> ```
> ERROR Error: Uncaught (in promise): ClientAuthError: URL navigated to is https://login.microsoftonline.com/<TenantId>/oauth2/v2.0/authorize?response_type=id_token&scope=openid%20profile&client_id=<ClientId>&redirect_uri=<RedirectUri>&state=<State>&nonce=<Nonce>&client_info=1&x-client-SKU=MSAL.JS&x-client-Ver=1.3.1&login_hint=<MyUsername>&client-request-id=<RequestId>&prompt=none&response_mode=fragment, Token renewal operation failed due to timeout.
> ```
> 
> Here's my code. There are 3 places where I'm configuring/calling MSAL. I don't _think_ I'm doing anything fancy. I'm basing it on the [Angular 8 sample app](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/angular8-sample-app/).
> 
> 1. `app.module.ts` -> `NgModule` -> `imports`:
>    ```
>    MsalModule.forRoot({
>            auth: {
>                clientId: "<ClientId>",
>                authority: "https://login.microsoftonline.com/<TenantId>/",
>                validateAuthority: true,
>                redirectUri: '<SiteRoot>',
>                postLogoutRedirectUri: "<SomeUrl>",
>                navigateToLoginRequestUrl: true,        
>            },
>            cache: {
>                cacheLocation : "localStorage",
>                storeAuthStateInCookie: isIE
>            },
>            framework: {
>                unprotectedResources: ['https://www.microsoft.com/en-us/'],
>            }
>        },
>        {
>            popUp: !isIE,
>            consentScopes: [ "https://graph.microsoft.com/User.Read", "api://<ApiId>/API" ],
>            extraQueryParameters: {},
>            protectedResourceMap: [
>                ['<SiteRoot>', ['api://<ApiId>/API']],
>                ['https://graph.microsoft.com', ['user.read', 'openid', 'profile', 'email', 'offline_access']]
>            ]
>        })
>    ```
> 2. `app.module.ts` -> `NgModule` -> `providers`
>    ```
>    { provide: HTTP_INTERCEPTORS, useClass: MsalInterceptor, multi: true }
>    ```
> 3. `app.component.ts` -> `AppComponent` -> `ngOnInit`
>    ```
>    ngOnInit() {
>        this.authService.handleRedirectCallback((authError, response) => {
>            if (authError) {
>                console.error('Redirect Error: ', authError.errorMessage);
>                return;
>            }
>    
>            console.log('Redirect Success: ', response);
>        });
>    }
>    ```


**This helped, the 3rd is where callback forces the redirect.** Easy to test in chrome using dev console. Clear cache, and click some link in the app, the MsalGuard would not redirect without given solution

![image](https://user-images.githubusercontent.com/2981715/99597765-19e69900-29be-11eb-800b-5977c034c12b.png)
